### ISSUE TITLE
Validate certificates

### ISSUE THREAD

**Issue by grampelberg** [linkerd/linkerd2#3431]
## What problem are you trying to solve?

It is way too easy to build the certificates wrong. Either because of a bad algorithm, misformating the actual data or time drift between generation and installation.

### How should the problem be solved?

Add a `check` that validates the cert can be used and has been generated correctly from the control plane secret itself.

**Comment by olix0r** [comment]
just to be clear on a detail: the validation should be done in the control plane itself (rather than reading the secrets into the CLI). we want to test the control plane's clock, especially, against the certificates

**Comment by adleong** [comment]
The identity controller does some validation on startup and panics if the validation fails.  Is this what we want to surface through `linkerd check`?  Or some other validation?

**Comment by grampelberg** [comment]
I'd like to:

- make sure we can load the certificates.
- that the algorithm is one we can use (eg not RSA).
- the certificate is valid from a not-before perspective.
- the trust bundle is valid (eg we can actually validate generated certificates).

**Comment by olix0r** [comment]
@adleong 

> The identity controller does some validation on startup and panics if the validation fails. Is this what we want to surface through linkerd check? Or some other validation?

I think what we basically want is for the identity service to expose an API that returns its public signing cert metadata, if valid, and, if invalid, returns the reason that the signing credentials cannot be used. This would allow clients that consume this API to make their own determinations with to regard what the acceptible certificate lifetime may be.  It's important that _only_ the identity service interact directly with the signing credentials.

I think we'll want to modify the identity to fail with a `grpc-status: INTERNAL`  and a `grpc-message` like _The identity controller is configured with invalid credentials._

**Comment by zaharidichev** [comment]
@grampelberg @adleong So does the scope of this still consist of: 

- Expose endpoint in the identity service as @ olix0r  described
- Add a check that uses this endpoint

Also, at the moment (after the automatic cert reloading got merged), the identity controller will simply not use any invalid certificates that it has been configured with and will return `Unavailable` status  with a message of `cert issuer not ready yet`

**Comment by ihcsim** [comment]
Following our post standup discussion, we've decided to not expose a new endpoint in the `identity` service, because if the certificate is invalid, the `identity` service (and possibly, the rest of the control plane) may not come up. This also implies that we won't be able to reliably test the control plane clock. FWIW, the `linkerd check --pre` command already performs a pre-installation clock test. See https://github.com/linkerd/linkerd2/pull/2803.

The going forward solution is to let the CLI fetch the trust root and the issuer key/cert pair directly. We chatted briefly about splitting the certificate into a config map, but the private key is still needed to perform some of the validation.

(Note that `linkerd check` already checks for secret-related RBAC permissions. See https://github.com/linkerd/linkerd2/pull/3639.)

Then introduce the following certificate checks to the `install` and `check` commands:

1. Ensure the key is generated using the ECDSA algorithm
1. Check the certificate lifetime:
   * Report an error if it expired
   * Issue a warning if it will be expiring soon (say 60 days?)
1. Repeat the [`verifyCred()`](https://github.com/linkerd/linkerd2/blob/1c879ac430c1dff329db64ea8b9e0e3f38529dc1/cli/cmd/install.go#L1067-L1077) checks in the `linkerd check` command. 

Updating the `check` command is necessary to support Helm users.
